### PR TITLE
object-store: remove S3ConditionalPut::ETagPutIfNotExists

### PR DESCRIPTION
**Pull request by benesch**
Now that real S3 supports `If-Match`, we no longer need this special conditional put mode for real S3.

Considerations:

  * This PR is based on https://github.com/apache/arrow-rs/pull/6801. 
  * This PR can't be merged until we upgrade to a version of localstack which supports `If-Match` (which doesn't yet exist). @bentsku mentioned he'd try to add support to localstack later this week. See https://github.com/apache/arrow-rs/issues/6799#issuecomment-2499239456.
  * Tests are expected to fail until ☝🏽 is addressed.

# Which issue does this PR close?

Cleanup related to https://github.com/apache/arrow-rs/issues/6799.

# Are there any user-facing changes?

The removal of `S3ConditionalPut::ETagPutIfNotExists` would be user facing, but this variant has not yet made it into a release.

cc @tustvold

### CODE REVIEW

**Comment by bentsku** [comment]
Quick update on the LocalStack side: I've got the green light to release this feature during this week, and will work on it either tonight or tomorrow. The feature will be available in the `latest` docker image then (I'll ping you once it is so you can also maybe test it locally?), and on Friday we will release `localstack/localstack:4.0.3`, so you could use that in CI. Sorry to hold you off!

**Comment by tustvold** [comment]
That's amazing, thank you for working on this!

**Comment by benesch** [comment]
> Sorry to hold you off!

Quite the opposite—thanks for turning the implementation around so quickly!

**Comment by bentsku** [comment]
As written in https://github.com/apache/arrow-rs/issues/6799#issuecomment-2503803712, the feature is not available in the `latest` image tag, and will be available in tagged release (`4.0.3`) on Friday 29th (time still be decided, depending on what needs to get in). 

I've given a try to this branch with the flag enabled and it worked for me locally, might be worth trying it out. Thanks a lot for the tests, they're always great to run locally! 🙏

**Comment by benesch** [comment]
Yes, just bumped to the `latest` localstack image in this PR and CI passed with flying colors! 🎉 Thanks very much, @bentsku.

Give me a holler when 4.0.3 is released—I'll try to keep an eye on it myself—so that I can update this PR to use it. And then we should be ready to merge.

**Comment by bentsku** [comment]
Hello @benesch! The `4.0.3` image is now available! A bit later than initially planned, we were ironing some quirks in other AWS services, but things are all good now. 

Can't wait to see this feature merged! 🚀 thanks a lot for the collaboration on this. Have a great week-end 😄

File: .github/workflows/object_store.yml [hunk 1]
```diff
@@ -138,7 +138,7 @@ jobs:
 
       - name: Setup LocalStack (AWS emulation)
         run: |
-          echo "LOCALSTACK_CONTAINER=$(docker run -d -p 4566:4566 localstack/localstack:3.8.1)" >> $GITHUB_ENV
+          echo "LOCALSTACK_CONTAINER=$(docker run -d -p 4566:4566 localstack/localstack:latest)" >> $GITHUB_ENV
```
Comment (tustvold) [inline .github/workflows/object_store.yml:141]:
```suggestion
          echo "LOCALSTACK_CONTAINER=$(docker run -d -p 4566:4566 localstack/localstack:4.0.3)" >> $GITHUB_ENV
```

**Review by tustvold — APPROVED** [review]

**Pull request merged**